### ISSUE TITLE
numeric exceptions in simple functions

### ISSUE THREAD

**Issue by BishopWolf** [numpy/numpy#17466]
Numeric types shall be self contained and automatically typecasted

### Reproducing code example:

```python
import numpy as np
a = np.array([1, 1, 1, 1, 1])
a += 0.1
```

### expected behaviour:

a is typecasted to float64 automatically before doing the add function. The following code works:

```python
import numpy as np
a = np.array([1., 1., 1., 1., 1.])
a += 0.1
```

it returns `array([1.1, 1.1, 1.1, 1.1, 1.1])`. This shall be the returning value of the code above.

### Error message:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
numpy.core._exceptions.UFuncTypeError: Cannot cast ufunc 'add' output from dtype('float64') to dtype('int64') with casting rule 'same_kind'
```

### NumPy/Python version information:

1.19.1 3.8.5 (default, Jul 21 2020, 10:48:26) 
[Clang 11.0.3 (clang-1103.0.32.62)]

**Comment by charris** [comment]
Doesn't work that way. The inplace operators in numpy reuse the same memory, so the result type must remain the same.

**Comment by BishopWolf** [comment]
this works
```
>>> b=1
>>> b+=0.1
>>> b
1.1
```

**Comment by charris** [comment]
It is also not "inplace". It is implemented as `b = b + 0.1`, which is completely different.
